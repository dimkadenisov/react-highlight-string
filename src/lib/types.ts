import { GetChunksType } from 'highlight-string'
import { ComponentType, CSSProperties, HTMLAttributes, ReactHTML } from 'react'

type WithActiveIndex = {
  /**
   * The class name to be applied to an active match. Use along with activeIndex
   */
  activeClassName?: string
  /**
   * Specify the match index that should be actively highlighted. Use along with activeClassName
   */
  activeIndex: number
  /**
   * The inline style to be applied to an active match. Use along with activeIndex
   */
  activeStyle?: CSSProperties
}

type WithoutActiveIndex = {
  activeClassName?: never
  activeIndex?: never
  activeStyle?: never
}

export type HighlightComponentProps = {
  activeIndex: number
  highlightIndex: number
  /**
   * CSS class or classes that will contain `activeClassName` and `highlightClassName` if they were provided
   */
  className?: string
} & HTMLAttributes<Element>

export type HighlighterProps = GetChunksType &
  (WithActiveIndex | WithoutActiveIndex) & {
    /**
     * CSS class name applied to the outer/wrapper span
     */
    className?: string
    /**
     * 	CSS class name applied to highlighted text or object mapping search term matches to class names.
     */
    highlightClassName?: string
    /**
     * Inline styles applied to highlighted text
     */
    highlightStyle?: CSSProperties
    /**
     * Type of tag to wrap around highlighted matches. Defaults to mark but can also be a React component (class or functional) */
    highlightComponent?: keyof ReactHTML | ComponentType<HighlightComponentProps>
    /**
     * CSS class name applied to unhighlighted text
     */
    unhighlightClassName?: string
    /**
     * Inline styles applied to unhighlighted text
     */
    unhighlightStyle?: CSSProperties
    /**
     * 	Type of tag applied to unhighlighted parts. Defaults to React Fragment but can also be a React component (class or functional)
     */
    unhighlightComponent?: keyof ReactHTML | ComponentType<HTMLAttributes<Element>>
  } & HTMLAttributes<HTMLSpanElement>
