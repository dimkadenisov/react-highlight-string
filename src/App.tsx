import './index.css'

import { useState } from 'react'
import { HighlightComponentProps, Highlighter } from 'react-highlight-string'

function HighlightComponent({ children, ...restProps }: HighlightComponentProps) {
  return <b {...restProps}>{children}</b>
}

const components = [
  <Highlighter text="Hello word" substrings={['hello']} />,
  <Highlighter text="Hello word" substrings={['hello']} caseSensitive />,
  <Highlighter text="Hello word" substrings={['hello', 'w']} />,
  <Highlighter text="Hello word" substrings={['hello', 'w']} caseSensitive />,
  <Highlighter text="Hello word" substrings={['hello']} className="highlighter" />,
  <Highlighter
    text="Hello word"
    substrings={['hello']}
    highlightComponent={HighlightComponent}
    unhighlightComponent={({ children, ...props }) => <i {...props}>{children}</i>}
  />,
]

function App() {
  const [activeIndex, setActiveIndex] = useState(0)

  return (
    <div className="layout">
      {components.map((component, index) => (
        <div className="layout_item" key={index}>
          {component}
        </div>
      ))}
      <div className="layout_item">
        <button
          onClick={() => {
            setActiveIndex((prev) => Math.max(prev - 1, 0))
          }}
        >
          (-) Decrease active index
        </button>
        <button
          onClick={() => {
            setActiveIndex((prev) => Math.min(prev + 1, 3))
          }}
        >
          (+) Increase active index
        </button>
        <br />
        <br />
        <div>
          <span>Active Index:</span> {activeIndex}
        </div>
        <br />
        <Highlighter
          text="Somewhere over the rainbow Bluebirds fly And the dreams that you dream of Dreams really do come true-ooh-ooh Someday I'll wish upon a star Wake up where the clouds are far behind me Where trouble melts like lemon drops High above the chimney tops that's where You'll find me, oh"
          substrings={['the']}
          className="highlighter"
          activeIndex={activeIndex}
          activeClassName="active"
        />
      </div>
    </div>
  )
}

export default App
