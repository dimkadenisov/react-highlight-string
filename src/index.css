:root {
  --body-bg: #f5f5f5;
  --text-main: #171717;
  --text-highlight: #ff0000;
  --layout-item-bg: #dcdcdc;
}

@media (prefers-color-scheme: dark) {
  :root {
    --body-bg: #202124;
    --text-main: #e8eaed;
    --layout-item-bg: #565656;
  }
}

body {
  background-color: var(--body-bg);
  color: var(--text-main);
  font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
}

.layout {
  display: flex;
  flex-direction: column;
  row-gap: 12px;
  max-width: 460px;
  margin: 60px auto;
}

.layout_item {
  padding: 12px;
  border-radius: 4px;
  background-color: var(--layout-item-bg);
}

.active {
  background-color: var(--text-highlight);
}
