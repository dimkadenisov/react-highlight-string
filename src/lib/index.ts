export { Highlighter } from './Highlighter'
export type { HighlightComponentProps, HighlighterProps } from './types'
