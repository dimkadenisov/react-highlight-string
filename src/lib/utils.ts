export function mergeClassNames(...classNames: (string | null | undefined)[]): string {
  return classNames.filter(Boolean).join(' ')
}
