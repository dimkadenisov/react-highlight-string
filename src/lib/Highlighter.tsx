import { getChunks } from 'highlight-string'

import { HighlighterProps } from './types'
import { mergeClassNames } from './utils'

export function Highlighter({
  text,
  substrings,
  caseSensitive,
  sanitize,
  activeClassName,
  activeIndex,
  activeStyle,
  highlightClassName,
  highlightStyle,
  highlightComponent: HighlightComponent = 'mark',
  unhighlightClassName,
  unhighlightStyle,
  unhighlightComponent: UnhighlightComponent = 'span',
  ...htmlProps
}: HighlighterProps) {
  const chunks = getChunks({
    text,
    substrings,
    caseSensitive,
    sanitize,
  })

  let highlightIndex = -1

  return (
    <span {...htmlProps}>
      {chunks.map(({ text, highlighted }, index) => {
        if (highlighted) {
          highlightIndex++
          const isActive = activeIndex === highlightIndex

          return (
            // to avoid passing highlightIndex and activeIndex props to html components
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            <HighlightComponent
              className={
                isActive ? mergeClassNames(highlightClassName, activeClassName) : highlightClassName
              }
              style={isActive ? { ...highlightStyle, ...activeStyle } : highlightStyle}
              key={index}
              {...(typeof HighlightComponent !== 'string' && { highlightIndex, activeIndex })}
            >
              {text}
            </HighlightComponent>
          )
        }

        return (
          <UnhighlightComponent
            className={unhighlightClassName}
            style={unhighlightStyle}
            key={index}
          >
            {text}
          </UnhighlightComponent>
        )
      })}
    </span>
  )
}
