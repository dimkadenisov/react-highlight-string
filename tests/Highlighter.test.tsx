import React from 'react'
import renderer from 'react-test-renderer'
import { expect, test } from 'vitest'

import { HighlightComponentProps, Highlighter } from '../src/lib'

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)

  return result as renderer.ReactTestRendererJSON
}

test('Simple test', () => {
  const component = renderer.create(<Highlighter text="Hello word" substrings={['hello']} />)

  const tree = toJson(component)

  expect(tree).toMatchSnapshot()
})

test('Case sensitive', () => {
  const component = renderer.create(
    <Highlighter text="Hello word" substrings={['hello']} caseSensitive />,
  )

  const tree = toJson(component)

  expect(tree).toMatchSnapshot()
})

test('Multiple substrings', () => {
  const component = renderer.create(<Highlighter text="Hello word" substrings={['hello', 'w']} />)

  const tree = toJson(component)

  expect(tree).toMatchSnapshot()
})

test('Multiple substrings case sensitive', () => {
  const component = renderer.create(
    <Highlighter text="Hello word" substrings={['hello', 'w']} caseSensitive />,
  )

  const tree = toJson(component)

  expect(tree).toMatchSnapshot()
})

test('With class name', () => {
  const component = renderer.create(
    <Highlighter text="Hello word" substrings={['hello']} className="highlighter" />,
  )

  const tree = toJson(component)

  expect(tree).toMatchSnapshot()
})

test('Active index and class name', () => {
  const component = renderer.create(
    <Highlighter
      text="Somewhere over the rainbow Bluebirds fly And the dreams that you dream of Dreams really do come true-ooh-ooh Someday I'll wish upon a star Wake up where the clouds are far behind me Where trouble melts like lemon drops High above the chimney tops that's where You'll find me, oh"
      substrings={['the']}
      className="highlighter"
      activeIndex={0}
      activeClassName="active"
    />,
  )
  const tree = toJson(component)

  expect(tree).toMatchSnapshot()

  tree.activeIndex = 1

  expect(tree).toMatchSnapshot()

  tree.activeIndex = 2

  expect(tree).toMatchSnapshot()

  tree.activeIndex = 3

  expect(tree).toMatchSnapshot()

  tree.activeIndex = 4

  expect(tree).toMatchSnapshot()
})

function HighlightComponent({ children, highlightIndex, className }: HighlightComponentProps) {
  return <mark className={`${className} ${highlightIndex % 2 ? 'odd' : 'even'}`}>{children}</mark>
}

test('Highlight component', () => {
  const component = renderer.create(
    <Highlighter
      text="Hello word"
      substrings={['hello']}
      className="highlighter"
      highlightComponent={HighlightComponent}
      unhighlightComponent={({ children, ...props }) => <i {...props}>{children}</i>}
    />,
  )

  const tree = toJson(component)

  expect(tree).toMatchSnapshot()
})
