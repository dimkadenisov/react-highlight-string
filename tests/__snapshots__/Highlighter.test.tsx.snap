// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Active index and class name 1`] = `
<span
  className="highlighter"
>
  <span>
    Somewhere over 
  </span>
  <mark
    className="active"
    style={{}}
  >
    the
  </mark>
  <span>
     rainbow Bluebirds fly And 
  </span>
  <mark>
    the
  </mark>
  <span>
     dreams that you dream of Dreams really do come true-ooh-ooh Someday I'll wish upon a star Wake up where 
  </span>
  <mark>
    the
  </mark>
  <span>
     clouds are far behind me Where trouble melts like lemon drops High above 
  </span>
  <mark>
    the
  </mark>
  <span>
     chimney tops that's where You'll find me, oh
  </span>
</span>
`;

exports[`Active index and class name 2`] = `
<span
  className="highlighter"
>
  <span>
    Somewhere over 
  </span>
  <mark
    className="active"
    style={{}}
  >
    the
  </mark>
  <span>
     rainbow Bluebirds fly And 
  </span>
  <mark>
    the
  </mark>
  <span>
     dreams that you dream of Dreams really do come true-ooh-ooh Someday I'll wish upon a star Wake up where 
  </span>
  <mark>
    the
  </mark>
  <span>
     clouds are far behind me Where trouble melts like lemon drops High above 
  </span>
  <mark>
    the
  </mark>
  <span>
     chimney tops that's where You'll find me, oh
  </span>
</span>
`;

exports[`Active index and class name 3`] = `
<span
  className="highlighter"
>
  <span>
    Somewhere over 
  </span>
  <mark
    className="active"
    style={{}}
  >
    the
  </mark>
  <span>
     rainbow Bluebirds fly And 
  </span>
  <mark>
    the
  </mark>
  <span>
     dreams that you dream of Dreams really do come true-ooh-ooh Someday I'll wish upon a star Wake up where 
  </span>
  <mark>
    the
  </mark>
  <span>
     clouds are far behind me Where trouble melts like lemon drops High above 
  </span>
  <mark>
    the
  </mark>
  <span>
     chimney tops that's where You'll find me, oh
  </span>
</span>
`;

exports[`Active index and class name 4`] = `
<span
  className="highlighter"
>
  <span>
    Somewhere over 
  </span>
  <mark
    className="active"
    style={{}}
  >
    the
  </mark>
  <span>
     rainbow Bluebirds fly And 
  </span>
  <mark>
    the
  </mark>
  <span>
     dreams that you dream of Dreams really do come true-ooh-ooh Someday I'll wish upon a star Wake up where 
  </span>
  <mark>
    the
  </mark>
  <span>
     clouds are far behind me Where trouble melts like lemon drops High above 
  </span>
  <mark>
    the
  </mark>
  <span>
     chimney tops that's where You'll find me, oh
  </span>
</span>
`;

exports[`Active index and class name 5`] = `
<span
  className="highlighter"
>
  <span>
    Somewhere over 
  </span>
  <mark
    className="active"
    style={{}}
  >
    the
  </mark>
  <span>
     rainbow Bluebirds fly And 
  </span>
  <mark>
    the
  </mark>
  <span>
     dreams that you dream of Dreams really do come true-ooh-ooh Someday I'll wish upon a star Wake up where 
  </span>
  <mark>
    the
  </mark>
  <span>
     clouds are far behind me Where trouble melts like lemon drops High above 
  </span>
  <mark>
    the
  </mark>
  <span>
     chimney tops that's where You'll find me, oh
  </span>
</span>
`;

exports[`Case sensitive 1`] = `
<span>
  <span>
    Hello word
  </span>
</span>
`;

exports[`Highlight component 1`] = `
<span
  className="highlighter"
>
  <mark
    className="undefined even"
  >
    Hello
  </mark>
  <i>
     word
  </i>
</span>
`;

exports[`Multiple substrings 1`] = `
<span>
  <mark>
    Hello
  </mark>
  <span>
     
  </span>
  <mark>
    w
  </mark>
  <span>
    ord
  </span>
</span>
`;

exports[`Multiple substrings case sensitive 1`] = `
<span>
  <span>
    Hello 
  </span>
  <mark>
    w
  </mark>
  <span>
    ord
  </span>
</span>
`;

exports[`Simple test 1`] = `
<span>
  <mark>
    Hello
  </mark>
  <span>
     word
  </span>
</span>
`;

exports[`With class name 1`] = `
<span
  className="highlighter"
>
  <mark>
    Hello
  </mark>
  <span>
     word
  </span>
</span>
`;
